controllers/operatorpolicy_controller.go

```diff
@@ -60,166 +70,211 @@ var _ reconcile.Reconciler = &OperatorPolicyReconciler{}
 
+			err := r.DynamicWatcher.EndQueryBatch(watcher)
+			if err != nil {
+				OpLog.Error(err, "Error ending the query batch in progress. Ignoring the failure.")
+			}
+
+			err = r.DynamicWatcher.RemoveWatcher(watcher)
+			if err != nil {
+				OpLog.Error(err, "Error updating dependency watcher. Ignoring the failure.")
+			}
+
 			return reconcile.Result{}, nil
 		}
+
+		OpLog.Error(err, "Failed to get operator policy")
+
+		return reconcile.Result{}, err
 	}
 
-	return reconcile.Result{}, nil
-}
+	// handle the policy
+	OpLog.Info("Reconciling OperatorPolicy", "policy", policy.Name)
 
-// PeriodicallyExecOperatorPolicies loops through all operatorpolicies in the target namespace and triggers
-// template handling for each one. This function drives all the work the operator policy controller does.
-func (r *OperatorPolicyReconciler) PeriodicallyExecOperatorPolicies(freq uint, elected <-chan struct{},
-) {
-	OpLog.Info("Waiting for leader election before periodically evaluating operator policies")
-	<-elected
+	// Check if specified Subscription exists
+	cachedItem, err := r.DynamicWatcher.Get(watcher, subscriptiongvk, policy.Spec.Subscription.Namespace,
+		policy.Spec.Subscription.SubscriptionSpec.Package)
+	if err != nil {
+		panic(err)
+	}
 
-	exiting := false
-	for !exiting {
-		start := time.Now()
+	subExists := cachedItem != nil
 
-		var skipLoop bool
-		if !skipLoop {
-			policiesList := policyv1beta1.OperatorPolicyList{}
+	// Check if any OperatorGroups exist in the namespace
+	ogNamespace := policy.Spec.Subscription.Namespace
+	if policy.Spec.OperatorGroup != nil {
+		ogNamespace = policy.Spec.OperatorGroup.Namespace
+	}
 
-			err := r.List(context.TODO(), &policiesList)
-			if err != nil {
-				OpLog.Error(err, "Failed to list the OperatorPolicy objects to evaluate")
-			} else {
-				for i := range policiesList.Items {
-					policy := policiesList.Items[i]
-					if !r.shouldEvaluatePolicy(&policy) {
-						continue
-					}
-
-					// handle policy
-					err := r.handleSinglePolicy(&policy)
-					if err != nil {
-						OpLog.Error(err, "Error while evaluating operator policy")
-					}
-				}
-			}
-		}
+	cachedList, err := r.DynamicWatcher.List(watcher, operatorgroupgvk, ogNamespace, nil)
+	if err != nil {
+		panic(err)
+	}
 
-		elapsed := time.Since(start).Seconds()
-		if float64(freq) > elapsed {
-			remainingSleep := float64(freq) - elapsed
-			sleepTime := time.Duration(remainingSleep) * time.Second
-			OpLog.V(2).Info("Sleeping before reprocessing the operator policies", "seconds", sleepTime)
-			time.Sleep(sleepTime)
-		}
+	ogExists := len(cachedList) != 0
+
+	// Exists indicates if a Subscription or Operatorgroup need to be created
+	exists := subExists && ogExists
+	shouldExist := strings.EqualFold(string(policy.Spec.ComplianceType), string(policyv1.MustHave))
+
+	// Case 1: policy has just been applied and related resources have yet to be created
+	if !exists && shouldExist {
+		return r.createPolicyResources(policy, watcher)
```

Comment:
Could you address @JustinKuli's comment about deferring `EndQueryBatch`? I find it hard to follow if the function starting the query batch isn't the one responsible for ending it.